sql/execinfra: don't adapt row.Fetcher to RowSource

andreimatei opened the pull request:
Before this patch, a row.Fetcher could be consumed as a RowSource
through RowFetcherWrapper. This was unnecessary and confusing; the
Fetcher did not quite match that interface. It has its own Start()
methods different from the RowSource's Start (and so now both had to be
called) making clients awkward. Clients were also failing to treat the
fetcher as a RowSource in other ways to: some of them were not draining
it properly. Which is cool, given that there's nothing to drain from a
fetcher. But it was confusing and also the error returned from the
Fetcher were handled awkwardly - wrapped in a meta.
Seemingly the reason why the RowFetcherWrapper was introduced was so
that we can have a common way to collect stats on top of either a
Fetcher or a RowSource (enter RowFetcherStatCollector).
RowFetcherStatsCollector was a very confusing beast since it aliased its
input, referencing it in two ways and only Start()ing one of the
aliases.
Instead I've separated a "RowSource+stats" wrapper from a separate
"Fetcher+stats" wrapper, resulting in a better life.

This change is good on its own, but I've made it particuarly since I'm
trying to change the RowSource interface to take a Txn at Start() time,
which would have made it truly confusing for the fetchers that would
have ended up with multiple Start methods looking now pretty similar.

Release note: None

andreimatei committed:
rowexec: use a TableReader's ctx for the underlying Fetcher

Before this patch, the fetcher under a table reader was using a ctx
different than the TableReader's - a ctx without the TR's log tag. I
apparently thought that's a good idea back in the day for some reason.
But now I think there's no reason for it. Also, it's in my way. So I'm
making the fetcher simply use the TR's ctx.

Release note: None

andreimatei committed:
sql/execinfra: don't adapt row.Fetcher to RowSource

Before this patch, a row.Fetcher could be consumed as a RowSource
through RowFetcherWrapper. This was unnecessary and confusing; the
Fetcher did not quite match that interface. It has its own Start()
methods different from the RowSource's Start (and so now both had to be
called) making clients awkward. Clients were also failing to treat the
fetcher as a RowSource in other ways to: some of them were not draining
it properly. Which is cool, given that there's nothing to drain from a
fetcher. But it was confusing and also the error returned from the
Fetcher were handled awkwardly - wrapped in a meta.
Seemingly the reason why the RowFetcherWrapper was introduced was so
that we can have a common way to collect stats on top of either a
Fetcher or a RowSource (enter RowFetcherStatCollector).
RowFetcherStatsCollector was a very confusing beast since it aliased its
input, referencing it in two ways and only Start()ing one of the
aliases.
Instead I've separated a "RowSource+stats" wrapper from a separate
"Fetcher+stats" wrapper, resulting in a better life.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/40897)
<!-- Reviewable:end -->

andreimatei commented:
I'm not gonna merge this on its own; I've got bigger plans that need this to address https://github.com/cockroachdb/cockroach/issues/40487

But it can be reviewed on its own.

yuzefovich reviewed (approved):
Thanks! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but I'd like Alfonso to take a look as well.

> Reviewed 1 of 1 files at r1, 7 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40897#-:-Lp61wWeDUEg7vEOzHYL:bn5jobx)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei and @asubiotto)

___
*[pkg/sql/execinfra/stats.go, line 138 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40897#-Lp61Xbs47oQsrOELF7c:-Lp61Xbs47oQsrOELF7d:blb4guc) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c030e0677283cdea2c9b97a5c91dfe78dc63c1/pkg/sql/execinfra/stats.go#L138)):*
> ```Go
> type RowFetcherStatCollector struct {
> 	*RowFetcherWrapper
> 	inputStatCollector *InputStatCollector
> ```

I think now `InputStatCollector` can be removed as well.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40897#-:-Lp7hd4cAACVSgLWLaTS:brfqlux)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @andreimatei and @asubiotto)



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40897)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/execinfra/stats.go, line 138 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40897#-Lp61Xbs47oQsrOELF7c:-Lp9ltFv2O8vlRPmvWQ1:befa4uq) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c030e0677283cdea2c9b97a5c91dfe78dc63c1/pkg/sql/execinfra/stats.go#L138)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think now `InputStatCollector` can be removed as well.
</blockquote></details>

InputStatCollector is still useful for wrapping a `RowSource`. It's used by a bunch of processors (separately from the processors that have a `RowFetcher`).

Or did you have a deeper thought?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40897)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/sql/execinfra/stats.go, line 138 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40897#-Lp61Xbs47oQsrOELF7c:-Lp9nakc3yxbFHEqOl26:b-jd5o2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/93c030e0677283cdea2c9b97a5c91dfe78dc63c1/pkg/sql/execinfra/stats.go#L138)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

InputStatCollector is still useful for wrapping a `RowSource`. It's used by a bunch of processors (separately from the processors that have a `RowFetcher`).

Or did you have a deeper thought?
</blockquote></details>

No, I briefly looked into where it was used, and it seemed to me that `RowFetcher` was the only user. If not, please disregard my comment.


<!-- Sent from Reviewable.io -->

andreimatei committed:
execinfra: make InputStatsCollector implement OpNode

We were using InputStatsCollectors as inputs to various OpNodes when the
query was traced, but these collectors weren't implementing OpNode. This
is illegal, and led to crashes when EXPLAIN(vec) was visiting the OpNode
tree.

Fixes #41011

Release note: Fix a crash when EXPLAIN(vec)-ing some index joins.

andreimatei reviewed (commented):
Piled on another small commit. PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40897#-:-LpZS1bt8pAgQTGvuneu:bstqmk2)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andreimatei and @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
> Reviewed 6 of 7 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40897)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andreimatei)

___
*[pkg/sql/execinfra/stats.go, line 49 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40897#-LpZTkUI7vNpSldR-2K3:-LpZTkUI7vNpSldR-2K4:b-fkxqsa) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9b91e653c8f8509046466064d5457a3e1310493/pkg/sql/execinfra/stats.go#L49)):*
> ```Go
> // Child is part of the OpNode interface.
> func (isc *InputStatCollector) Child(nth int) OpNode {
> 	return isc.RowSource.(OpNode)
> ```

[nit]: as a sanity check, in other implementations of this method we verify that `nth` is in bounds and panic if it's not.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40897#-:-Lql7GaV2KohnenBaBOs:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andreimatei and @yuzefovich)

___
*[pkg/sql/execinfra/stats.go, line 49 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/40897#-LpZTkUI7vNpSldR-2K3:-Lql5aIQ4fBX2U3rL4b2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9b91e653c8f8509046466064d5457a3e1310493/pkg/sql/execinfra/stats.go#L49)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: as a sanity check, in other implementations of this method we verify that `nth` is in bounds and panic if it's not.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
